### DISCUSSION TITLE
Choco bindings for high-level programming langages (e.g. R, Python)

### DISCUSSION THREAD

**Discussion by dimitri-justeau** [chocoteam/choco-solver#835] [Ideas]
I would love to see people from other communities having fun with Choco (e.g. biologists, climatologists, ...). However, outside the computer science world, people mainly use high-level languages such as R or Python for their analysis. Of course, we have MiniZinc, which can be used from R and Python. But does MiniZinc with Choco as backend offer as much flexibility as Choco alone does? (I believe not, but I might be wrong).

Developing from scratch R or Python bindings for Choco would be an enormous task, but maybe there is a way to do this automatically, with an automatic documentation generation from the javadoc ? I am currently not aware of automatic binding tools from Java to other languages, but I know that it exists for C++: http://www.swig.org/ , https://jslee02.github.io/awesome-cpp-python-binding-generator/ , ...

**Comment by dimitri-justeau** [comment]
GraalVM might be helpful (see https://github.com/chocoteam/choco-solver/discussions/836) with its polyglot capabilities.

**Comment by cprudhom** [comment]
I agree with you, having higher-level languages or languages more accessible to non computer scientists is a good idea/opportunity.
Actually, [PyCSP3](https://pypi.org/project/pycsp3/) is a Python library to model and solve (with ACE and choco-solver as back-end solvers) CP models. There is also [CPMpy](https://pypi.org/project/cpmpy/), but it does not support choco-solver.
But, what you are looking for is other languages binders, which requires, as you noticed, times to achieve...
